### PR TITLE
fix(core): preserve multi-line caret navigation in ChatComposerInput

### PR DESCRIPTION
**Pull request by Lee-Dongwook**
## Related Issue
https://github.com/facebook/astryx/issues/4282

  ## Summary

  Fixes `ChatComposerInput` message-history navigation so ArrowUp and ArrowDown only recall history when the caret is at the relevant text
  boundary.

  This preserves the browser’s default vertical caret movement within multi-line drafts.

  ## Changes

  - Recall the previous message with ArrowUp only when the selection is at the start of the draft.
  - Move forward through message history with ArrowDown only when the selection is at the end of the draft.
  - Preserve repeated history navigation when a recalled message is fully selected.
  - Restore the pending draft after navigating past the newest history entry.
  - Add selection-boundary helpers for contenteditable composer input.
  - Add a patch changeset for `@astryxdesign/core`.

  ## Tests

  Added coverage for:

  - ArrowUp recall at the start of a draft
  - Default ArrowUp behavior when the caret is mid-text, including multi-line content
  - Forward/backward history navigation
  - Pending-draft restoration
  - Empty history and `hasHistory={false}`

### CODE REVIEW

**Comment by cixzhang** [comment]
Merged current `main` into this branch (it was 544 commits behind) so CI runs against something current — no code changes. Prettier is clean and the `ChatComposerInput` tests pass locally (58/58).

File: packages/core/src/Chat/chatComposerSelection.ts [hunk 1]
```diff
@@ -57,6 +57,78 @@ export function ensureCaretInside(editable: HTMLElement): Selection | null {
   return selection;
 }
 
+/**
+ * Whether a range spans no visible content — no text and no element
+ * (tokens, `<br>`). An empty or whitespace-free range between a
+ * boundary and the caret means the caret is at that boundary.
+ */
+function isRangeEmpty(range: Range): boolean {
+  const contents = range.cloneContents();
```
Comment (cixzhang) [inline packages/core/src/Chat/chatComposerSelection.ts]: This copies the full draft on every boundary arrow key. Might need a non-copying boundary check.

**Review by cixzhang — CHANGES_REQUESTED** [review]
Thanks — the caret and history behavior is right in Chromium. One thing before this lands: both boundary helpers clone the editor contents, so an ArrowUp/Down at a boundary now copies the entire draft before the existing history work. Could we avoid content-sized copying in this keydown path and bank a count-based regression?

If you'd rather talk it through with someone, we're in [Discord](https://discord.com/invite/XnsUcFykEP).

[Reviewed by Robohands]

**Comment by Lee-Dongwook** [comment]
> Thanks — the caret and history behavior is right in Chromium. One thing before this lands: both boundary helpers clone the editor contents, so an ArrowUp/Down at a boundary now copies the entire draft before the existing history work. Could we avoid content-sized copying in this keydown path and bank a count-based regression?
> 
> If you'd rather talk it through with someone, we're in [Discord](https://discord.com/invite/XnsUcFykEP).
> 
> [Reviewed by Robohands]
> 

Thanks for answering. Since I watched your's comment now, I'll re-check by myself in this PR with  following your's feedback!

**Comment by Lee-Dongwook** [comment]
@cixzhang 

Hello! Here's the change summary of commit.

---

Addressed the cloneContents feedback in eec64c5a5.

- Replaced the content-cloning boundary checks with Range.compareBoundaryPoints(), so ArrowUp/ArrowDown boundary detection does not create a DocumentFragment proportional to the draft size.
- Added count-based regression coverage for both ArrowUp and ArrowDown boundary paths, asserting that Range.prototype.cloneContents is called zero times.
- Kept the existing behavior for collapsed carets and fully selected recalled history entries.

Validation: git diff --check passed. I could not run Vitest locally because dependencies are not installed in this checkout and the package registry DNS lookup is unavailable.

File: packages/core/src/Chat/chatComposerSelection.ts (hunk 1 above)
Comment (Lee-Dongwook) [reply packages/core/src/Chat/chatComposerSelection.ts]: Removed the existing `isRangeEmpty()` helper and its `range.cloneContents()` call. The caret’s start/end position is now compared directly with the editor boundaries using `Range.compareBoundaryPoints()`. As a result, ArrowUp/ArrowDown no longer creates a DocumentFragment by copying the entire draft.

**Review by cixzhang — CHANGES_REQUESTED** [review]
Thanks — removing the content clone fixes the prior cost, but the boundary check now misses the positions Chromium actually creates. At the start of a non-empty draft Chromium reports `#text` offset 0; `compareBoundaryPoints` returns 1, so ArrowUp leaves the draft unchanged. The tests were changed to place selection on the editor root, which is why they pass. Please cover real text-node start/end positions without content-sized copying.

If you'd rather talk it through with someone, we're in [Discord](https://discord.com/invite/XnsUcFykEP).

[Reviewed by Robohands]

**Comment by Lee-Dongwook** [comment]
Thanks for catching this. I replaced the `compareBoundaryPoints` check with a DOM-edge walk that recognizes Chromium’s actual caret positions: offset 0 in the first text node and text.length in the final text node.
  
The check does not clone or materialize draft content; it only inspects the boundary offset and relevant sibling/ancestor paths. I also updated the `ArrowUp` and `ArrowDown` regression tests to place the selection on real text-node boundaries and kept assertions that `Range.prototype.cloneContents` is never called.
  
git diff --check passed.

**Review by cixzhang — APPROVED** [review]
Thanks, this is good. My earlier boundary checks now pass in real Chromium: multiline ArrowUp stays in the draft, text-node boundaries recall history, and ArrowDown restores it.

[Reviewed by Robohands]
